controller/src/main/java/io/pravega/controller/server/ControllerService.java

```diff
@@ -509,5 +510,48 @@ private void reportDeleteStreamMetrics(String scope, String streamName, DeleteSt
         }
     }
 
+    public CompletableFuture<Controller.TimestampResponse> noteTimestampFromWriter(String scope, String stream, String writerId, long timestamp, Map<Long, Long> streamCut) {
+        return streamStore.noteWriterMark(scope, stream, writerId, timestamp, streamCut, null, executor)
+                .handle((r, e) -> {
+                    Controller.TimestampResponse.Builder response = Controller.TimestampResponse.newBuilder();
+                    if (e != null) {
+                        response.setResult(Controller.TimestampResponse.Status.INTERNAL_ERROR);
+                    } else {
+                        switch (r) {
+                            case SUCCESS:
+                                response.setResult(Controller.TimestampResponse.Status.SUCCESS);
+                                break;
+                            case INVALID_TIME:
+                                response.setResult(Controller.TimestampResponse.Status.INVALID_TIME);
+                                break;
+                            case INVALID_POSITION:
+                                response.setResult(Controller.TimestampResponse.Status.INVALID_POSITION);
+                                break;
+                            default:
+                                response.setResult(Controller.TimestampResponse.Status.INTERNAL_ERROR);
+                                break;
+                        }
+                    }
+                    return response.build();
+                });
+    }
+ 
+    public CompletableFuture<Controller.WriterShutdownResponse> shutdownWriter(String scope, String stream, String writer) {
+        return streamStore.removeWriter(scope, stream, writer, null, executor)
```

Comment:
Shutdown writer is the grpc call. Its connotation is to tell controller to stop performing any action on behalf of the said writer wrt watermarking. 
The way controller makes use of writers is to store their marks durably in the store. And then a periodic workflow scans the store to retrieve latest marks stored against all known writers and then compute and emit a watermark (if any). 
So to shutdown computation for a writer really is about removing the writer from the underlying store so that we do not consider it for computation. Hence remove writer.

Reply:
In fact, i just realized, we should have both shutdown and remove semantics on the store. 
This is because of the following:
if a writer reports a writer mark, its expectation is that this mark will be included in the watermark computation.
But watermark computation happens asynchronously. 
So if the writer shutdown is called before the asynchronous computation is performed, and we remove the writer mark, then we do not include the writer's mark in the computation. 
We should mark the writer for shutdown and it should be removed from computation if it becomes inactive and not before that.

Reply:
There are two parts to this discussion:

*Method name*
`remove` is a much more natural term to refer to what we are doing here. We have a group of writers for which we are collecting watermarks. One element of the group is being decommissioned and needs to be removed. Whether that element is going to be shutdown or not is being the scope of this call. I think we should not use `shutdown` here.

*Semantics of removal*
It is correct that the asynchronous nature of the computation is such that operations can happen out of order. But, instead of creating another call to fix the problem with semantics, we should instead serialize the operations in the case the property is really needed. I'm not in favor of a second call, it seems unnecessary and confusing.

Reply:
> Semantics of removal
> It is correct that the asynchronous nature of the computation is such that operations can happen out of order. But, instead of creating another call to fix the problem with semantics, we should instead serialize the operations in the case the property is really needed. I'm not in favor of a second call, it seems unnecessary and confusing.

You misunderstood perhaps. The writers report their marks. Once they have reported it, their expectation would be that it should be included in watermarking computation. 
They could then call shutdown/remove. If we remove the writer mark before its included in the watermark computation then we break the expectation. 
Since watermark computation is asynchronous and periodic, the way i have implemented the logic is as follows (this is in the followup PR and you can refer to the draft PR open here: #3839 ). 
When a writer requests for a shutdown, Controller simply updates a flag inside the writer's mark that it has been shutdown. 
During watermarking computation, after having consumed the writer's mark, if the writer has been marked for shutdown, the workflow also removes the writer. 

So shutdown and remove have separate semantics from metadata store's perspective. 
From application's perspective, they have the notion of `shutdown` of a writer. We could rename it to some other verb like remove

Reply:
I believe I understood the problem right, but perhaps did not express my thought correctly. I understood that there can be a race in the case a writer notes a mark followed by removing itself. Due to asynchronous and unordered nature of the computation, it is possible that the removal is processed before the mark is noted. What I am saying is that the way to solve it is to make sure the operations are linearized. Even if they are executed asynchronously, they can still following submission order.

The part I have probably misunderstood is your proposal. Reading again, you say:

> i just realized, we should have both shutdown and remove semantics on the store.

which I initially interpreted as a separate API call altogether, put perhaps you meant to say that this is a way to avoid the race.

> From application's perspective, they have the notion of shutdown of a writer. We could rename it to some other verb like remove

The application might be closing a client, and the client needs to remove itself from this note group. Any synonym of "remove" will do for me. :-)

Reply:
So Remove Writer and Watermark computation are linearized in my proposal. 

Call from caller merely marks the writer for shutdown. 
The removal is performed by next watermark computation after having consumed the writer's last reported mark. 

So for name change -- I could rename the grpc api from `ShutdownWriter` to `CloseWriter`.